firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -156,53 +156,47 @@ String getName() {
   @WorkerThread
   private String getFirebaseInstallationId() throws FirebaseInstallationsException {
 
-    PersistedFidEntryValue persistedFidEntryValue = persistedFid.readPersistedFidEntryValue();
+    PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
 
-    if (persistedFidDoesNotExistsOrInErrorState(persistedFidEntryValue)) {
-      return createFidAndPersistFidEntry();
+    if (persistedFidExistsAndRegistered(persistedFidEntry)) {
+      return persistedFidEntry.getFirebaseInstallationId();
     }
 
-    if (persistedFidExistsAndRegistered(persistedFidEntryValue)) {
-      return persistedFidEntryValue.getFirebaseInstallationId();
+    if (persistedFidMissingOrInErrorState(persistedFidEntry)) {
+      return createFidAndPersistFidEntry();
     }
 
-    if (persistedFidExistsAndUnregistered(persistedFidEntryValue)) {
-      registerAndSaveFid(persistedFidEntryValue.getFirebaseInstallationId());
-      return persistedFidEntryValue.getFirebaseInstallationId();
+    if (persistedFidExistsAndUnregistered(persistedFidEntry)) {
+      Tasks.call(executor, () -> registerAndSaveFid(persistedFidEntry.getFirebaseInstallationId()));
+      return persistedFidEntry.getFirebaseInstallationId();
     }
 
     return null;
```

Comment:
How can this method ever return NULL?

Reply:
If delete() is called and there is no network availability, we mark registration status as deleted.  In that case, it returns NULL.

Is that not right?

Reply:
@andirayo Can you please check this? I think the issue is, we have introduced a state DELETED as per fis-android-sdk-design doc. But from your comments on my doc, seems like you don't want the DELETED state. 

We can change the above function as :

  @WorkerThread
  private String getFirebaseInstallationId() throws FirebaseInstallationsException {
    PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();

    if (persistedFidExistsAndRegistered(persistedFidEntry)) {
      return persistedFidEntry.getFirebaseInstallationId();
    } else  if (persistedFidExistsAndUnregistered(persistedFidEntry)) {
      Tasks.call(executor, () -> registerAndSaveFid(persistedFidEntry.getFirebaseInstallationId()));
      return persistedFidEntry.getFirebaseInstallationId();
    }
      return createFidAndPersistFidEntry();
    }

Wdyt?

Reply:
Except for the "else if" in there (an else is not necessary after you "return" in the if-statement, I like it better yes.

How about
// not registered with FIS yet
if (fidExistsAndUnregistered(persistedFidEntry)) {
  Tasks.call(executor, () -> registerAndSaveFid(persistedFidEntry.getFirebaseInstallationId()));
}
// return FID from in local storage
if (fidExists(persistedFidEntry)) {
  return persistedFidEntry.getFirebaseInstallationId();
}

return createFidAndPersistFidEntry();

Reply:
https://help.github.com/en/articles/creating-and-highlighting-code-blocks

Reply:
what do you all think about something like this? I haven't checked that it'll compile, but it seems to clearly express what we're trying to do:

1. return the existing fid if it's there
2. create one if it didn't exist
3. trigger a background registration if it wasn't previously registered

```java
public Task<FID> getFid() {
  return getExistingFid()
    .continueWith(orElse(this::createAndPersistNewFid))
    .onSuccessTask(this::registerFidIfNecessary);
}

public static <F, T> Continuation<F, T> orElse(Supplier<T> supplier) {
  return t -> {
    if (t.isSuccessful()) { return t.getResult(); }
    return supplier.get();
  };
}

private FID createAndPersistNewFid() {
  FID fid = Utils.createRandomFid();
  persist(fid);
  return fid;
}

private void registerFidIfNecessary(FID fid) {
  // do network stuff
}
```

Reply:
@ciarand I tried incorporating the changes you mentioned below:

```
  @RequiresApi(api = Build.VERSION_CODES.N)
  private Task<String> getFid() {
    return Tasks.call(executor, this::getExistingFid)
        .continueWith(orElse(new NewFidSupplier()))
        .onSuccessTask(this::registerFidIfNecessary);
  }

  private PersistedFidEntry getExistingFid() throws FirebaseInstallationsException {
    PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
    if (persistedFidMissingOrInErrorState(persistedFidEntry)) {
      throw new FirebaseInstallationsException(
          "Failed to get existing fid.", FirebaseInstallationsException.Status.CLIENT_ERROR);
    }
    return persistedFidEntry;
  }

  @RequiresApi(api = Build.VERSION_CODES.N)
  @NonNull
  public static <F, T> Continuation<F, T> orElse(@NonNull Supplier<T> supplier) {
    return t -> {
      if (t.isSuccessful()) {
        return (T) t.getResult();
      }
      return supplier.get();
    };
  }

  @RequiresApi(api = Build.VERSION_CODES.N)
  class NewFidSupplier implements Supplier<PersistedFidEntry> {

    @Override
    public PersistedFidEntry get() {
      return createAndPersistNewFid();
    }
  }

  private PersistedFidEntry createAndPersistNewFid() {
    String fid = Utils.createRandomFid();
    try {
      persistFid(fid);
      PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
      return persistedFidEntry;
    } catch (FirebaseInstallationsException e) {
      // Exception handling
    }
    return PersistedFidEntry.builder().build();
  }

  private void persistFid(String fid) throws FirebaseInstallationsException {
    boolean firstUpdateCacheResult =
        persistedFid.insertOrUpdatePersistedFidEntry(
            PersistedFidEntry.builder()
                .setFirebaseInstallationId(fid)
                .setRegistrationStatus(PersistedFid.RegistrationStatus.UNREGISTERED)
                .build());

    if (!firstUpdateCacheResult) {
      throw new FirebaseInstallationsException(
          "Failed to update client side cache.",
          FirebaseInstallationsException.Status.CLIENT_ERROR);
    }
  }

  private Task<String> registerFidIfNecessary(PersistedFidEntry persistedFidEntry) {
    if (persistedFidEntry.getRegistrationStatus() != PersistedFid.RegistrationStatus.REGISTERED) {
      Tasks.call(executor, () -> registerAndSaveFid(persistedFidEntry));
    }
    return Tasks.forResult(persistedFidEntry.getFirebaseInstallationId());
  }

  /** Registers the created Fid with FIS Servers and update the shared prefs. */
  private Void registerAndSaveFid(PersistedFidEntry persistedFidEntry)
      throws FirebaseInstallationsException {
    // network call
  }
```

Reply:
I have 2 concerns: 

- Supplier get call requires API level 24, Is this right way of handling it? 
- Exception handling by the Supplier while persisting the newly created fid

Reply:
Yeah we can't use the standard java.util.functional.Supplier interface. We can just define one in the parent class:

```
interface Supplier<T> {
  T get() throws Exception;
}
```

When you say "Exception handling by the Supplier while persisting the newly created fid" do you mean you're worried that you can't propagate exceptions or you're worried that they will be propagated?